lib/publiccloud/instance.pm

```diff
@@ -349,56 +351,68 @@ sub wait_for_ssh {
+    $self->ssh_script_run("ssh-keyscan $args{public_ip} | tee -a ~/.ssh/known_hosts")
+      if (get_var('PUBLIC_CLOUD_PERF_COLLECT') or get_var('PUBLIC_CLOUD_CHECK_BOOT_TIME'));
+
+    my $out;
+    if ($status) {
+        $out = "ssh access OK";
     }
-
-    script_run("ssh  -i /root/.ssh/id_rsa $args{username}\@$args{public_ip} true", timeout => 360);
-    # Debug output: We have occasional error in 'journalctl -b' - see poo#96464 - this will be removed soon.
-    # Exclude 'mr_test/saptune' test case as it will introduce random softreboot failures.
-    if (!get_var('PUBLIC_CLOUD_SLES4SAP')) {
-        $self->run_ssh_command(cmd => 'sudo journalctl -b', proceed_on_failure => 1, username => $args{username}, timeout => 360);
+    else {
+        $out = "ssh access KO" if ($args{proceed_on_failure});
     }
-
-    unless ($args{proceed_on_failure}) {
-        my $error_msg;
-        if ($check_port) {
-            $error_msg = sprintf("Unable to reach SSH port of instance %s with public IP:%s within %d seconds", $self->{instance_id}, $self->{public_ip},
-                $args{timeout});
+    # check also remote system is running, looping by --wait option:
+    if ($args{systemup_check} and $status) {
+        # timeout recalculated removing consumed time until now
+        $out = $self->ssh_script_output(cmd => 'sudo systemctl is-system-running --wait',
+            timeout => $args{timeout} - $duration, proceed_on_failure => 1, username => $args{username});
+
+        $duration = time() - $start_time;    # time update after check result
+        if ($out =~ m/running/) {    # startup OK
+            $status = 1;
+        }
+        elsif ($out =~ m/degraded/) {    # OK with failed services: collection
+            $status = 1;
+            $out .= "\n" . $self->ssh_script_output(cmd => 'sudo systemctl --failed', proceed_on_failure => 1, username => $args{username});
+        }
+        else {    # FAIL: $out ~ maintenance|stopping|offline|unknown or else
+            $status = 0;
+            $out .= "\n" . "Can not reach systemd target on " . $instance_msg;
         }
-        else {
-            $error_msg = sprintf("Can not reach systemd target on instance %s with public IP:%s within %d seconds",
-                $self->{instance_id}, $self->{public_ip}, $args{timeout});
+        # Log upload
+        if (!get_var('PUBLIC_CLOUD_SLES4SAP')) {
+            #Exclude 'mr_test/saptune' test case as it will introduce random softreboot failures.
+            $self->ssh_script_run('sudo journalctl -b --no-pager > /tmp/journalctl.log',
+                timeout => 360, proceed_on_failure => 1, username => $args{username});
+            $self->upload_log('/tmp/journalctl.log', failok => 1);
         }
-        croak($error_msg);
     }
-
-    return;
+    # result display
+    record_info("CHECK " . ($args{systemup_check} ? "SYSTEM" : "SSH"), "Duration: $duration sec.\nResult: $out");
```

Comment:
Add
```perl
croak('Booting timeout exceeded!') if ($duration > $args{timeout});
```

Reply:
Already performed in `script_retry` by *fail_msg*

Reply:
why ? isn't this will fail execution ? what about `$args{proceed_on_failure}` in such case ?

Reply:
That information is already present in the ending `croak` call, at/near L420, and the proper message will be composed inside the previous `if` branches.